File: Graph/Java/Max Area of Island.java
// Problem link: https://leetcode.com/problems/max-area-of-island/description/

// Approach: 1 (DFS)
// Time: O(n * m)
// Space: O(n * m)
class Solution {
    public int maxAreaOfIsland(int[][] grid) {
        int rows = grid.length, cols = grid[0].length, result = 0;
        boolean[][] visit = new boolean[rows][cols];
        int[][] directions = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

        for(int i = 0; i < rows; i++)
            for(int j = 0; j < cols; j++)
                result = Math.max(result, this.dfs(i, j, grid, visit, directions));

        return result;
    }

    private int dfs(int row, int col, int[][] grid, boolean[][] visit, int[][] directions) {
        if(row < 0 || col < 0 || row == grid.length || col == grid[0].length || grid[row][col] == 0 || visit[row][col])
            return 0;
        
        visit[row][col] = true;

        int count = 1;
        for(int[] dir : directions)
            count = count + this.dfs(dir[0] + row, dir[1] + col, grid, visit, directions);

        return count;
    }
}


// Approach: 2 (BFS)
// Time: O(n * m)
// Space: O(n * m)
class Solution {
    public int maxAreaOfIsland(int[][] grid) {
        int rows = grid.length, cols = grid[0].length, result = 0;
        boolean[][] visit = new boolean[rows][cols];
        int[][] directions = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

        for(int i = 0; i < rows; i++)
            for(int j = 0; j < cols; j++)
                if(grid[i][j] == 1 && !visit[i][j])
                    result = Math.max(result, this.bfs(i, j, grid, visit, directions));
        
        return result;
    }

    private int bfs(int row, int col, int[][] grid, boolean[][] visit, int[][] directions) {
        Queue<int[]> queue = new LinkedList<>();
        int rows = grid.length, cols = grid[0].length, count = 0, r, c;
        queue.offer(new int[] {row, col});
        visit[row][col] = true;
        int[] cell;

        while(!queue.isEmpty()) {
            cell = queue.poll();
            count++;

            for(int[] dir : directions) {
                r = dir[0] + cell[0];
                c = dir[1] + cell[1];
                if(r >= 0 && r < rows && c >= 0 && c < cols && !visit[r][c] && grid[r][c] == 1) {
                    queue.offer(new int[] {r, c});
                    visit[r][c] = true;
                }
            }
        }

        return count;
    }
}

File: Graph/Python/Max Area of Island.py
# Problem link: https://leetcode.com/problems/max-area-of-island/description/

# Approach: 1 (DFS)
# Time: O(n * m)
# Space: O(n * m)
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        if not grid or not grid[0]:
            return 0
        
        rows, cols, result = len(grid), len(grid[0]), 0
        visit = set()
        directions = [[1, 0], [0, 1], [-1, 0], [0, -1]]
        result = 0

        for i in range(rows):
            for j in range(cols):
                result = max(result, self.dfs(i, j, grid, visit, directions))
        
        return result
        
    def dfs(self, row, col, grid, visit, directions):
        if row < 0 or col < 0 or row >= len(grid) or col >= len(grid[0]) or grid[row][col] == 0 or (row, col) in visit:
            return 0
            
        visit.add((row, col))
        count = 1

        for dr, dc in directions:
            count += self.dfs(row + dr, col + dc, grid, visit, directions)
        
        return count



# Approach: 2 (BFS)
# Time: O(n * m)
# Space: O(n * m)
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        if not grid or not grid[0]:
            return 0
        
        rows, cols = len(grid), len(grid[0])
        visit = set()
        directions = [(1, 0), (0, 1), (-1, 0), (0, -1)]
        result = 0
        
        for i in range(rows):
            for j in range(cols):
                if grid[i][j] == 1 and (i, j) not in visit:
                    result = max(result, self.bfs(i, j, grid, visit, directions))
        
        return result
    
    def bfs(self, row, col, grid, visit, directions):
        queue = deque([(row, col)])
        visit.add((row, col))
        count = 0
        
        while queue:
            r, c = queue.popleft()
            count += 1
            
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                if nr >= 0 and nr < len(grid) and nc >= 0 and nc < len(grid[0]) and (nr, nc) not in visit and grid[nr][nc] == 1:
                    queue.append((nr, nc))
                    visit.add((nr, nc))
        
        return count
